### PR TITLE
CP-34028: implement Uuid with Uuidm

### PR DESCRIPTION
**Pull request by psafont**
The Uuid module is problematic because it doesn't parse input at all. This PR tries to do a better job at that and replaces its base type from string to Uuidm, which should makes the representation more compact when used at the expense of conversion roundtrips when reading the database

In order to guard from invalid values it's necessary to change the interface to use Option: using exceptions would be dangerous and some uses might be missed and lead to unhandled exceptions.

Draft PR as it has a breaking change so clients need to be modified first; I would like some feedback before doing that.

TODO: The functions that generate random UUID do not use the ones provided by Uuidm. I don't think the Option.get is worrying because previously a different exception would have been raised, but there's a chance to make it more robust.

Notes:
- While having the phantom type makes sense to avoid comparing uuid of different types of resources, xapi needs to have a dedicated type to model uuids instead of using strings.

- On top of this, using Uuid.t means that for every roundtrip we're paying the cost of converting from the string representation to the raw byte value and viceversa which may impact performance, having the string representation with the new interface may be more desirable.

Original PR can be found in https://github.com/xapi-project/xen-api-libs-transitional/pull/103

### CODE REVIEW

**Comment by lindig** [comment]
I think this is a good change but it touches a lot of code. Random thoughts:

* The old code supported  UUIDs of different length and we always use the `V4` variant. I assume we don't need the flexibility to change this and `Uuid` as a module does not need to be as general.

* Most use cases need a UUID in string representation. This could be supported directly with a `make_string` or similar function to avoid `Uuid.to_string (Uuid.make ())` everywhere.

* Given that we mostly use strings, where is the compact representation used?

File: ocaml/libs/uuid/uuid.mli [hunk 1]
```diff
@@ -12,52 +12,66 @@
...
-(** Create a fresh UUID *)
+val null : 'a t
+(** A null UUID, as if such a thing actually existed. It turns out to be
+    useful though. *)
 
-val make_uuid_prng : unit -> 'a t
+val make : unit -> 'a t
+(** Create a fresh UUID *)
 
 val make_uuid_urnd : unit -> 'a t
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/libs/uuid/uuid.mli:36]: I'll remove the make_uuid_urnd and document the random source for make.

**Comment by psafont** [comment]
> The old code supported UUIDs of different length and we always use the V4 variant. I assume we don't need the flexibility to change this and Uuid as a module does not need to be as general.

This is the first time I hear UUIDs can have variable length, RFC  4122 is clear on this:
>   The UUID format is 16 octets; some bits of the eight octet variant field specified below determine finer structure.

Is there a non-standard source of UUIDs that we have to support? I'll test the changes before undrafting anyway to try to discover any unpleasant surprises

> Most use cases need a UUID in string representation. This could be supported directly with a make_string or similar function to avoid Uuid.to_string (Uuid.make ()) everywhere.
Yeah, this makes sense

> Given that we mostly use strings, where is the compact representation used?

I would like to use the compact representation more often and push strings only for storing to the database, it has the added benefit of specializing the type to a class of objects, e.g. `[ `Vm] Uuid.t` so they cannot be mixed up.

**Comment by edwintorok** [comment]
We have to be careful here to keep using a secure way of generating UUIDs, because the session UUID is used as an access token in API calls once a password login has succeeded.
The code here uses `/dev/urandom` which should be safe enough (except perhaps during early boot).

**Review by edwintorok — APPROVED** [review]
Can you fix the confclits? What remains to be done to get this ready?

**Comment by psafont** [comment]
I think now it's much better than before, it needs a build for end-to-end testing

**Review by lindig — APPROVED** [review]

**Pull request merged**